cranelift/assembler-x64/meta/src/generate/inst.rs

```diff
@@ -228,83 +229,194 @@ impl dsl::Inst {
     }
 
-    /// `(decl x64_<inst> (<params>) <return>)
-    ///  (extern constructor x64_<inst> x64_<inst>)`
+    /// `(decl x64_<inst>_raw (<params>) <return>)`
+    /// `(decl x64_<inst> (<params>) <return>)`
+    /// `(decl x64_<inst>_mem (<params>) <return>)` (maybe)
     ///
     /// # Panics
     ///
     /// This function panics if the instruction has no operands.
     pub fn generate_isle_definition(&self, f: &mut Formatter) {
-        use dsl::OperandKind::*;
-
+        // First declare the "raw" constructor which is implemented in Rust
+        // with `generate_isle_macro` above. This is an "extern" constructor
+        // with relatively raw types. This is not intended to be used by
+        // general lowering rules in ISLE.
         let struct_name = self.name();
-        let rule_name = format!("x64_{struct_name}");
+        let raw_name = format!("x64_{struct_name}_raw");
         let params = self
             .format
             .operands
             .iter()
-            .filter_map(|o| match o.location.kind() {
-                FixedReg(_) => None,
-                Imm(loc) => {
-                    let bits = loc.bits();
-                    if o.extension.is_sign_extended() {
-                        Some(format!("AssemblerSimm{bits}"))
-                    } else {
-                        Some(format!("AssemblerImm{bits}"))
-                    }
-                }
-                Reg(_) => Some(format!("Assembler{}Gpr", o.mutability.generate_type())),
-                RegMem(_) => Some(format!("Assembler{}GprMem", o.mutability.generate_type())),
-            })
+            .filter(|o| o.mutability.is_read())
+            // FIXME(#10238) don't filter out fixed regs here
+            .filter(|o| !matches!(o.location.kind(), OperandKind::FixedReg(_)))
+            .collect::<Vec<_>>();
+        let raw_param_tys = params
+            .iter()
+            .map(|o| o.isle_param_raw())
             .collect::<Vec<_>>()
             .join(" ");
-        let ret = match self.format.operands.first().unwrap().location.kind() {
-            Imm(_) => unreachable!(),
-            FixedReg(_) | Reg(_) => "AssemblerReadWriteGpr",
-            RegMem(_) => "AssemblerReadWriteGprMem",
-        };
+        f.line(format!("(decl {raw_name} ({raw_param_tys}) AssemblerOutputs)"), None);
+        f.line(format!("(extern constructor {raw_name} {raw_name})"), None);
 
-        f.line(format!("(decl {rule_name} ({params}) {ret})"), None);
-        f.line(format!("(extern constructor {rule_name} {rule_name})"), None);
+        // Next, for each "real" ISLE constructor being generated, synthesize a
+        // pure-ISLE constructor which delegates appropriately to the `*_raw`
+        // constructor above.
+        //
+        // The main purpose of these constructors is to have faithful type
+        // signatures for the SSA nature of VCode/ISLE, effectively translating
+        // x64's type system to ISLE/VCode's type system.
```

Comment:
Can we add an example here, just to see at a glance what the constructors look like? It's a little hard to pick out from the format string below otherwise.

Reply:
Like inline in the code comment or in this PR?

Reply:
In the code comment, yep, similar to how elsewhere there are some examples of `(decl ...)` and the like.

Reply:
Now that I'm looking at this, there are some examples up at the top of this function already like we've done elsewhere... But they are a bit sparse &mdash; I'll add a bit more detail.